include/exec/__detail/__system_context_default_impl_entry.hpp

```diff
@@ -23,23 +23,111 @@
+        (void) get_system_context_instance();
+      }
+
+      ~__system_context_keep_alive() {
+        release_system_context_instance(__system_context_instance.load());
+      }
+    };
+
+    struct __system_context_base {
+      __system_context_base() {
+        // This static will create the system context, and decrement its
+        // ref count after main exits.
+        static __system_context_keep_alive __s_keep_alive;
+      }
+    };
+  } // namespace __detail
+
+  /// Gets the default system context implementation.
+  extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE system_context_interface* get_system_context_instance() {
+    // spin until we get the lock:
+    __detail::__system_context_mutex.__lock();
+    // load the instance
+    auto* __instance = __detail::__system_context_instance.load();
+    // if it is not null, increment the reference count and return the ptr
+    if (__instance != nullptr) {
+      ++__instance->ref_count;
+      __detail::__system_context_mutex.__unlock();
+      return __instance;
+    }
+    __detail::__system_context_mutex.__unlock();
+
+    // create a new instance:
+    auto* __new_instance = __detail::__new_system_context_handler.load()();
+
+    // spin until we get the lock:
+    __detail::__system_context_mutex.__lock();
+
+    // try to set the new instance. if this fails, we have been usurped.
+    // increment the ref count on the usurper, destroy the instance we
+    // just created, and return the usurper.
+    if (!__detail::__system_context_instance.compare_exchange_strong(__instance, __new_instance)) {
+      ++__instance->ref_count;
+      __detail::__system_context_mutex.__unlock();
+      __new_instance->destroy_fn(__new_instance);
+      return __instance;
+    }
+
+    __detail::__system_context_mutex.__unlock();
+    return __new_instance;
+  }
+
+  /// Releases a reference to the specified system context object.
+  /// If the ref count drops to zero, then
+  ///   - if the specified object is the active global object, set
+  ///     the active global object to null.
+  ///   - destroy the instance
+  extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE void
+    release_system_context_instance(system_context_interface* __instance) noexcept {
+    // spin until we get the lock:
+    __detail::__system_context_mutex.__lock();
+    if (0 != --__instance->ref_count) {
+      __detail::__system_context_mutex.__unlock();
+      return;
+    }
+
+    // if we have just released the last reference on the global system
+    // context, set the global pointer to null:
+    auto* __instance_copy = __instance;
+    (void) __detail::__system_context_instance.compare_exchange_strong(__instance_copy, nullptr);
+
+    __detail::__system_context_mutex.__unlock();
+    __instance->destroy_fn(__instance);
+  }
+
+  /// Sets the default system context implementation.
+  extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE new_system_context_handler
+    set_new_system_context_handler(new_system_context_handler __handler) {
```

Comment:
Actually, I noticed that instead you have the ref-count initialized to 1 instead of to 0.

This instead means that nothing will ever destroy the system context object as nothing ever releases the last ref-count (assuming that `system_context` objects have matching calls to `get_system_context_instance()` and `release_system_context_instance()`).

Reply:
That is what keep_alive is for. It releases the final ref count.

Reply:
If it is the intent of the keep_alive to release the final ref-count then I don't think that the keep_alive constructor should be trying to increment the ref-count by calling `get_system_context_instance()` as you are already initializing the ref-count to 1.

However, it's hard to evaluate exactly where keep_alive constructor is being invoked as that code doesn't seem to be called from anywhere at the moment.

Reply:
`system_context` is meant to inherit from `__system_context_base`, so that the when the system context object is first created, its ref-count is `2` with the second ref-count held by the keep-alive until after main.